File: api/src/main/java/io/akitect/cms/model/Post.java
package io.akitect.cms.model;

import io.akitect.cms.model.base.BaseEntity;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "posts")
@Getter
@Setter
public class Post extends BaseEntity {

    @Column(name = "title", length = 255, nullable = false)
    private String title;

    @Column(name = "slug", length = 300, nullable = false, unique = true)
    private String slug;

    @Column(name = "content", columnDefinition = "text", nullable = false)
    private String content;

    @Column(name = "excerpt", columnDefinition = "text")
    private String excerpt;

    @Column(name = "featured_image_url", length = 255)
    private String featuredImageUrl;

    @Column(name = "status", length = 20, nullable = false)
    private String status = "DRAFT";

    @Column(name = "published_at")
    private LocalDateTime publishedAt;

    @Column(name = "meta_title", length = 255)
    private String metaTitle;

    @Column(name = "meta_description", columnDefinition = "text")
    private String metaDescription;

    @Column(name = "is_featured", nullable = false)
    private boolean featured = false;

    @Column(name = "allow_comments", nullable = false)
    private boolean allowComments = true;

    @Column(name = "views_count", nullable = false)
    private int viewsCount = 0;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    @ManyToMany
    @JoinTable(
            name = "post_categories",
            joinColumns = @JoinColumn(name = "post_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private Set<Category> categories = new HashSet<>();

    @ManyToMany
    @JoinTable(
            name = "post_tags",
            joinColumns = @JoinColumn(name = "post_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    private Set<Tag> tags = new HashSet<>();

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL)
    private Set<PostRevision> revisions = new HashSet<>();

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL)
    private Set<Comment> comments = new HashSet<>();
}

File: admin-panel/src/types/post.ts
export interface Category {
  id: string;
  name: string;
  slug: string;
  description?: string;
}

export interface Tag {
  id: string;
  name: string;
  slug: string;
}

export interface Post {
  id: string;
  title: string;
  slug: string;
  content: string;
  excerpt?: string;
  featuredImageUrl?: string;
  status: string;
  publishedAt?: string;
  metaTitle?: string;
  metaDescription?: string;
  featured: boolean;
  allowComments: boolean;
  viewsCount: number;
  createdAt: string;
  updatedAt: string;
  authorId: string;
  authorUsername: string;
  authorFullName?: string;
  categories?: Category[];
  tags?: Tag[];
}

export interface PostFilterParams {
  page?: number;
  size?: number;
  search?: string;
  status?: string;
  categoryId?: string;
  authorId?: string;
}

export interface PostCreateRequest {
  title: string;
  slug?: string;
  content: string;
  excerpt?: string;
  featuredImageUrl?: string;
  status: string;
  metaTitle?: string;
  metaDescription?: string;
  featured?: boolean;
  allowComments?: boolean;
  authorId?: string;
  categoryIds?: string[];
  tagIds?: string[];
}

export interface PostUpdateRequest {
  title: string;
  slug?: string;
  content: string;
  excerpt?: string;
  featuredImageUrl?: string;
  status: string;
  metaTitle?: string;
  metaDescription?: string;
  featured?: boolean;
  allowComments?: boolean;
  categoryIds?: string[];
  tagIds?: string[];
}

export interface PostRevision {
  id: string;
  postId: string;
  revisionNumber: number;
  title: string;
  content: string;
  excerpt?: string;
  createdAt: string;
  createdById: string;
  createdByUsername: string;
  createdByFullName?: string;
}

export interface PostState {
  posts: Post[];
  totalItems: number;
  totalPages: number;
  currentPage: number;
  selectedPost: Post | null;
  loading: boolean;
  error: string | null;
}